subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_espi.c

```diff
@@ -4,103 +4,161 @@
+	ESPI_STATE_SENDING,
+	ESPI_STATE_COUNT,
+};
 
-struct ec_host_cmd_backend_espi_data {
-	struct k_sem handler_owns;
-	struct k_sem dev_owns;
-	uint32_t rx_buffer_len;
+struct ec_host_cmd_espi_ctx {
+	/* eSPI device instance */
+	const struct device *espi_dev;
+	/* Context for read operation */
+	struct ec_host_cmd_rx_ctx *rx_ctx;
+	/* Transmit buffer */
+	struct ec_host_cmd_tx_buf *tx;
+	/* eSPI callback */
 	struct espi_callback espi_cb;
-	uint8_t *espi_shm;
+	/* eSPI Host Command state */
+	enum ec_host_cmd_espi_state state;
 };
 
-static void ec_host_cmd_backend_espi_handler(const struct device *dev, struct espi_callback *cb,
-					    struct espi_event espi_evt)
+#define EC_HOST_CMD_ESPI_DEFINE(_name)                                                             \
+	static struct ec_host_cmd_espi_ctx _name##_hc_espi;                                        \
+	struct ec_host_cmd_backend _name = {                                                       \
+		.api = &ec_host_cmd_api,                                                           \
+		.ctx = (struct ec_host_cmd_espi_ctx *)&_name##_hc_espi,                            \
+	}
+
+static void espi_handler(const struct device *dev, struct espi_callback *cb,
+			 struct espi_event espi_evt)
 {
-	struct ec_host_cmd_backend_espi_data *data =
-		CONTAINER_OF(cb, struct ec_host_cmd_backend_espi_data, espi_cb);
+	struct ec_host_cmd_espi_ctx *hc_espi =
+		CONTAINER_OF(cb, struct ec_host_cmd_espi_ctx, espi_cb);
 	uint16_t event_type = (uint16_t)espi_evt.evt_details;
+	/* tx stores the shared memory buf pointer and size, so use it */
+	const struct ec_host_cmd_request_header *rx_header = hc_espi->tx->buf;
+	const size_t shared_size = hc_espi->tx->len_max;
+	const uint16_t rx_valid_data_size = rx_header->data_len + RX_HEADER_SIZE;
 
 	if (event_type != ESPI_PERIPHERAL_EC_HOST_CMD) {
 		return;
 	}
 
-	if (k_sem_take(&data->dev_owns, K_NO_WAIT) != 0) {
-		int res = EC_HOST_CMD_IN_PROGRESS;
-
-		espi_write_lpc_request(ESPI_DEVICE, ECUSTOM_HOST_CMD_SEND_RESULT, &res);
+	/* Make sure we've received a Host Command in a good state not to override buffers for
+	 * a Host Command that is currently being processed. There is a moment between sending
+	 * a response and setting state to ESPI_STATE_READY_TO_RECV when we can receive a new
+	 * host command, so accept the sending state as well.
+	 */
+	if (hc_espi->state != ESPI_STATE_READY_TO_RECV && hc_espi->state != ESPI_STATE_SENDING) {
+		LOG_ERR("Received HC in bad state");
 		return;
 	}
 
-	k_sem_give(&data->handler_owns);
+	/* Only support version 3 and make sure the number of bytes to copy is not
+	 * bigger than rx buf size or the shared memory size
+	 */
+	if (rx_header->prtcl_ver != 3 ||
+	    rx_valid_data_size > CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER ||
+	    rx_valid_data_size > shared_size) {
+		memcpy(hc_espi->rx_ctx->buf, (void *)rx_header, RX_HEADER_SIZE);
+		hc_espi->rx_ctx->len = RX_HEADER_SIZE;
+		hc_espi->state = ESPI_STATE_RX_BAD;
```

Comment:
You're not using this state anywhere else, you should just set the state to `ESPI_STATE_PROCESSING` since you're passing the request to the general handler anyway.

Reply:
Done.